src/storage/dictionary_encoding_tile.cpp

```diff
@@ -0,0 +1,186 @@
+				type::Type::GetTypeSize(type::TypeId::TINYINT),
+				schema.GetColumn(i).GetName(), true);
+			columns.push_back(encoded_column);
+			dict_encoded_columns.insert(i);
+		} else {
+			columns.push_back(schema.GetColumn(i));
+		}
+		offset += schema.GetLength(i);
+	}
+
+	// if nothing needs to be encoded
+	if (dict_encoded_columns.empty()) return;
+
+	delete[] data;
+	schema = catalog::Schema(columns);
+	tuple_length = schema.GetLength();
+	tile_size = num_tuple_slots * tuple_length;
+	data = new char[tile_size];
+}
+
+// delete data is enough
+DictEncodedTile::~DictEncodedTile(){}
+
+type::Value DictEncodedTile::GetValue(const oid_t tuple_offset, const oid_t column_id) {
+	if (dict_encoded_columns.count(column_id) > 0) {
+		// encoded
+		auto idx_val = Tile::GetValue(tuple_offset, column_id);
+		auto idx = idx_val.GetAs<uint8_t>();
+    return type::Value(element_array[idx]);
+	} else {
+		return Tile::GetValue(tuple_offset, column_id);
+	}
+}
+
+// assume offset is based on original schema, so get schema give original schema
+type::Value DictEncodedTile::GetValueFast(const oid_t tuple_offset, const size_t column_offset,
+                           const type::TypeId column_type,
+                           const bool is_inlined){
+  auto column_id = original_schema_offsets[column_offset];
+	auto curr_val = Tile::GetValueFast(tuple_offset, schema.GetOffset(column_id),
+	  column_type, is_inlined);
+	if (dict_encoded_columns.count(column_id) > 0) {
+		auto idx_val = Tile::GetValue(tuple_offset, column_id);
+		auto idx = idx_val.GetAs<uint8_t>();
+    return type::Value(element_array[idx]);
+	} else {
+		return curr_val;
+	}
+}
+
+//void DictEncodedTile::SetValue(const type::Value &value, const oid_t tuple_offset,
+//                const oid_t column_id){
+//
+//}
+
+void DictEncodedTile::DictEncode(Tile *tile) {
+	if (is_dict_encoded) return;
+	LOG_INFO("dictionary encode, database_id: %d, table_id: %d, tile_group_id: %d"
+				", tile_id: %d", database_id, table_id, tile_group_id, tile_id);
+
+	// use a 2d vector to save the data
+	std::vector<std::vector<type::Value>> new_data_vector(column_count);
+
+	for (oid_t i = 0; i < column_count; i++) {
+		// if it is inlined, no need to compress!
+		if (dict_encoded_columns.count(i) > 0) {
+			LOG_INFO("encoding column %s", schema.GetColumn(i).column_name.c_str());
+			// to for tuple offset
+			for (oid_t to = 0; to < num_tuple_slots; to++) {
+				type::Value curr_old_val = tile->GetValue(to, i);
+				type::Value curr_val;
+				if (curr_old_val.GetTypeId() == type::TypeId::VARBINARY) {
+					curr_val = type::ValueFactory::GetVarbinaryValue((const unsigned char*) curr_old_val.GetData(),
+					    curr_old_val.GetLength(), true);
+				} else {
+					curr_val = type::ValueFactory::GetVarcharValue(curr_old_val.GetData(), curr_old_val.GetLength(),
+							true);
+				}
+				LOG_INFO("%s", curr_val.GetInfo().c_str());
+				// assume the idx take 1 byte
+				char idx_data[1];
```

Comment:
Consider renaming this variable. It's not clear what "idx_data" is from the name. Maybe "encoded_indices" ?